### PR TITLE
[Debug] Allow throwing from __toString() with `return trigger_error($e, E_USER_ERROR);`

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | - |
| License | MIT |
| Doc PR | - |

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
When one does `return trigger_error($exception, E_USER_ERROR);` from a __toString() method,
it is possible to convert the call to make it do the same as `throw $exception;`, which is otherwise forbidden from happening in __toString in PHP.
This convention plays nice with any error handler (even the native one). And when a compliant error handler is used (like ours if this PR gets merged) it can be used to throw from __toString().

File: src/Symfony/Component/Debug/Tests/Fixtures/ToStringThrower.php [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    {
+        $this->exception = $e;
+    }
+
+    public function __toString()
+    {
+        try {
+            throw $this->exception;
+        } catch (\Exception $e) {
+            return user_error($e, E_USER_ERROR);
```
Comment (Tobion) [inline src/Symfony/Component/Debug/Tests/Fixtures/ToStringThrower.php]: we use trigger_error everywhere else

Comment (nicolas-grekas) [inline src/Symfony/Component/Debug/Tests/Fixtures/ToStringThrower.php]: This is why I use `user_error` here: to make sure we do not forget about this alias.

Comment (nicolas-grekas) [inline src/Symfony/Component/Debug/Tests/Fixtures/ToStringThrower.php]: Comment added about that

Comment (Tobion) [inline src/Symfony/Component/Debug/Tests/Fixtures/ToStringThrower.php]: But then there is not test for trigger_error is there?

Comment (nicolas-grekas) [inline src/Symfony/Component/Debug/Tests/Fixtures/ToStringThrower.php]: That's true...

**Comment by nicolas-grekas** [comment]
The full workaround is the following:
- `trigger_error($e, E_USER_ERROR);` triggers the error handler
- $e is casted to string automatically and given to the error handler as $message
- the error handler looks up for the original exception in its $context variable, searching for a matching index whose value is an exception whose string representation is exactly $message
- the matched $e exception is temporarily stored in a private $toStringException static property
- the return value of trigger_error is always `null`, so we exit __toString() by returning `null`
- which triggers an E_RECOVERABLE_ERROR, because this is the behavior of __toString() when a not-string value is returned
- which calls the error handler again
- that checks the $toStringException property and throws it (yes, it is allowed at this point).

On HHVM, things are much simpler, because throwing from __toString is allowed.
The behavior ends up being the same (and that's the reason why the provided test case passes on HHVM also).

When no error handler is plugged, PHP dies with a fatal error (as usual in this situation), and with a message full of details: the exception string contains its type+message+stack trace excerpt.

**Comment by Tobion** [comment]
> the return value of trigger_error is always null

It actually is a boolean

File: src/Symfony/Component/Debug/ErrorHandler.php [hunk 2]
```diff
@@ -377,7 +378,10 @@ public function handleError($type, $message, $file, $line, array $context, array
         }
 
         if ($throw) {
-            if (($this->scopedErrors & $type) && class_exists('Symfony\Component\Debug\Exception\ContextErrorException')) {
+            if (isset(self::$toStringException)) {
```
Comment (Tobion) [inline src/Symfony/Component/Debug/ErrorHandler.php]: normally we do `null !== self::$toStringException`

**Comment by nicolas-grekas** [comment]
> It actually is a boolean

thanks, updated

File: src/Symfony/Component/Debug/ErrorHandler.php (hunk 2 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/Debug/ErrorHandler.php]: updated

**Comment by Tobion** [comment]
How did you find out about this hack? Never seen it before. We also had to add stupid workarounds in PSR-7 StreamInterface and UriInterface since throwing an exception in toString is not allowed in PHP.

**Comment by nicolas-grekas** [comment]
I don't remember exactly if I discovered this hack by myself or not, but I used it a few years ago in my own (now history) framework (https://github.com/nicolas-grekas/Patchwork/commit/0f1f1fdbe40df2204af20776bd1dd0dfd44aa957)

**Comment by Tobion** [comment]
What if I just want to create an exception without catching it first. Something like

```
class Uri {
    public function __toString() {
        trigger_error(new \LogicException('Given URI components  cannot be represented as URI reference'), E_USER_ERROR);
    }
}
```

Or do you get a stack trance with a trigger_error anyway?

```
    public function __toString() {
        trigger_error('Given URI components  cannot be represented as URI reference', E_USER_ERROR);
    }
```

**Comment by nicolas-grekas** [comment]
If you do not follow the convention (which is the case in both your examples), then you will end up in the `if` after the `foreach`:
https://github.com/symfony/symfony/pull/15076/files#diff-e7d086f2c2086fdb8f108e74f11dac5bR429

which will handle it as uncaught exception, then kill the process (by returning false).
We have no better choice here since throwing would also kill the process with a fatal error.

**Comment by fabpot** [comment]
:+1:

**Comment by stof** [comment]
:+1:

**Pull request merged**